### DISCUSSION TITLE
How to parallel python function

### DISCUSSION THREAD

**Discussion by sun-rs** [PyO3/pyo3#2391] [Questions]
Just learning pyo3 to wrap some python function and run in rust. Assume I have 10 python function use same data source, could I make them do the calculation in parallelism by pyo3?

**Comment by davidhewitt** [comment]
It sounds like you're asking whether PyO3 can let you run multiple copies of a Python function in parallel. The answer, is that just like in Python, you would have to use subprocesses with each running their own interpreter (e.g. via `multiprocessing` module).

What PyO3 can do is let you write a single Python function which has a parallel implementation in Rust, such as [this example](https://github.com/PyO3/pyo3/blob/48690525e19b87818b59f99be83f1e0eb203c7d4/examples/word-count/src/lib.rs#L6).

If it's a long-running algorithm then you can also use `allow_threads` to release the GIL and allow other Python code to make progress at the same time.
